frontstage/controllers/party_controller.py

```diff
@@ -229,32 +230,112 @@ def get_respondent_enrolments(party_id):
+    # Creates a list of threads which will call functions to set the survey, case, party and collex responses
+    # in the cache_data.
+
+    threads = []
+
+    for survey_id in surveys_ids:
+        threads.append(CachingDataThreadCreation(get_survey, cache_data, app.config, survey_id))
+        threads.append(CachingDataThreadCreation(get_collex, cache_data, app.config, survey_id))
+
+    for business_id in business_ids:
+        threads.append(CachingDataThreadCreation(get_case, cache_data, app.config, business_id, tag))
+        threads.append(CachingDataThreadCreation(get_party, cache_data, app.config, business_id))
+
+    for thread in threads:
+        thread.start()
+    for thread in threads:
+        thread.join()
+
+
+def caching_data_for_collection_instrument(cache_data):
+    # This function creates a list of threads from the collection instrument id in the cache_data of the cases.
+    collection_instrument_ids = set()
+    threads = []
+    for business_id, cases in cache_data['cases'].items():
+        for case in cases:
+            collection_instrument_ids.add(case['collectionInstrumentId'])
+    for collection_instrument_id in collection_instrument_ids:
+        threads.append(CachingDataThreadCreation(get_collection_instrument, cache_data, app.config,
+                                                 collection_instrument_id))
+
+    for thread in threads:
+        thread.start()
+    for thread in threads:
+        thread.join()
+
+# This method has gone through a rewrite in an attempt to make the to-do page more performant. Before the rewrite, it
+# was making REST calls in a for loop and this was causing frontstage to time out for some respondents trying to access
+# their surveys. Some of the calls were repeating what it's already done so, we've decided to
+# cache the relevant data before it enters the main for loop in attempt to increase the performance. We've
+# used sets to make sure its not using any duplicate ids for the business, survey and collection instrument. With the
+# ids we then cache the data into a dictionary which can then be used later.
+
+
 def get_survey_list_details_for_party(party_id, tag, business_party_id, survey_id):
+    enrolment_data = get_respondent_enrolments(party_id)
+
+    # Gets the survey ids and business ids from the enrolment data that has been generated.
+
+    surveys_ids, business_ids = set_enrolment_data(enrolment_data)
+
+    # This is a dictionary that will store all of the data that is going to be cached instead of making multiple calls
+    # inside of the for loop for get_respondent_enrolments.
+
+    cache_data = {'surveys': dict(),
+                  'businesses': dict(),
+                  'collexes': dict(),
+                  'cases': dict(),
+                  'instrument': dict()}
+
+    # These two will call the services to get responses and cache the data for later use.
+
+    caching_data_for_survey_list(cache_data, surveys_ids, business_ids, tag)
```

Comment:
Both caching_xxx_survey_xx and caching_ci_xxx do a start and join. Therefore we get survey data first then we get ci data . Why not have them both start and then do a join afterwards so these run asynchronously ? (unless the data from one is needed for the other ? But I dont THINK it does)

Reply:
We need this as two functions because before we can get the collection instruments, we need the cases to be cached to get the collection instrument ids from them. It might be a good idea to change the name of the first function but I kept it as survey_list because it's getting the important responses for the todo page.